### ISSUE TITLE
remote ssh plugin (like vscode remote-ssh)

### ISSUE THREAD

**Issue by justinmk** [neovim/neovim#21635]
This is mostly just for tracking. Obviously everyone wants this, but it looks like work... 😏 

# Problem

vscode has "remote" tools (broadly referred to as `vscode-remote`, in particular `vscode remote-ssh`)

- https://code.visualstudio.com/docs/remote/remote-overview
- https://code.visualstudio.com/docs/remote/ssh

... that allow the user to

1. input a ssh URI (hostname + port)
     - or select from a list of hosts discovered from your local `~/.ssh/config`
2. vscode connects to the remote ssh endpoint using your local `~/.ssh` credentials
    - or prompts for password as needed
3. vscode starts a new local UI
4. vscode installs _another vscode_ on the remote machine automatically
    - it also installs your plugins, on the remote!
5. the new local vscode UI controls the remote vscode server, and you can use it to work on the remote machine very much like a local vscode.

# Ideaaaaaa

1. Implement a "universal installer" script (sh, powershell) that pulls from 
https://github.com/neovim/neovim/releases and installs the correct Nvim on the current machine
   ```
    curl https://neovim.io/install-hax && bash install-hax
   ```
    - does something like this already exist? 
        - https://github.com/rhysd/action-setup-vim
        - https://github.com/NTBBloodbath/nvenv
        - https://github.com/MordechaiHadad/bob
2. Implement an Nvim plugin that works like vscode remote-ssh (described above)
    - select ssh host from `~/.ssh/config`
    - installs Nvim on the remote (if suitable `nvim` version is not found on the remote)
    - starts remote Nvim server
    - configures ssh tunnel
    - starts local Nvim UI connected to the remote Nvim

**Comment by lewis6991** [comment]
Relates to https://github.com/chipsenkbeil/distant.nvim

**Comment by theHamsta** [comment]
More related ideas:
- virtually (or via scp) adding local rtp to the remote environment? Or maybe have a dedicated lua file for it that let's users spin up their preferred environment: a remote equivalent to `init.lua` when `init.lua` is not yet present on the remote.
- the possibility to manage such attachments and quickly switch between them

**Comment by ObserverOfTime** [comment]
Is it really necessary to include this in the core?
Remote files can already be edited through `sshfs` with a local editor.
Using a remote editor should be accomplished with a plugin instead.

**Comment by justinmk** [comment]
> Is it really necessary to include this in the core (if that's the plan)?

afaik, this should be possible as a shell script + plugin (which is the current proposal described above). 

doesn't need to be in core, but until a plugin magically appears, this issue just tracks the general idea. And if a plugin does magically appear, then it might make sense to integrate it as a builtin plugin later.

**Comment by jdrouhard** [comment]
One major difference with vscode-remote vs a remote nvim UI is that with vscode-remote, all the keyboard input (and thus the actual buffer changes) are done by local client, with the input and server-driven updates being synchronized asynchronously in the background.

This becomes very important when you have a high latency connection--vscode-remote stays feeling "snappy" when typing with characters being displayed to the screen instantly, but a remote nvim will start to be laggy with displaying input (or even just a normal nvim run in an ssh session).

Not saying we should try to make it work like vscode-remote, but just wanted to point out the high latency remote connection use case as food for thought.

**Comment by justinmk** [comment]
yep, see also emacs [TRAMP](https://www.emacswiki.org/emacs/TrampMode). It's definitely a better model in many respects, but more complex. vscode has a [VFS](https://github.com/microsoft/vscode/blob/be54d2be91cc7443d0c44ed2a483475b12747b2d/src/vs/platform/workspace/common/virtualWorkspace.ts) layer, Nvim would need something like that.

The door is open for some sort of VFS model in Nvim, but that's orthogonal to this tracking issue.

**Comment by theHamsta** [comment]
What you can do in VS code is also to manage multiple remote sessions. Maybe there could be a command to reattach the TUI client to another server during a running session. Then, it could be implemented via a plugin only in a running session.

**Comment by chipsenkbeil** [comment]
@justinmk curious if there would be a way to embed the distant library into neovim itself. It's a rust library, so if we exported it in a C-friendly way, it could be consumed.

If so, you could have neovim provide tramp natively and support doing this via ssh and distant's own protocol. And anything else added eventually like Google Drive, Dropbox, etc. for file editing.

**Comment by meicale** [comment]
This is going to be a killer feature!

**Comment by KFearsoff** [comment]
Here's my 2 cents.

I would love to see remote development as a feature (perhaps in a plugin, but I do want it!), but I think following the footsteps of VSCode is the wrong approach. 

# VSCode approach downsides
1. Running Neovim on the remote takes CPU, RAM and disk space. This can be a complete dealbreaker for some of the more resource-intensive plugins, like `rust-analyzer`.
2. As noted by @jdrouhard , latencies quickly become an issue, which makes remote development feeling no better than just SSHing into the host and typing manually.
3. We'd have to write a script for installing Neovim. We want it to be reliable enough to not warrant any manual intervention. This is a **tall** order. Consider the sheer varierty of targets: you have Linux targets that use `glibc`, `musl`, ancient kernels, different BSDs, Windows, mingw, MacOS. All of them with their own quirks. Differences in Neovim dependencies might introduce subtle bugs that are basically impossible to debug.
4. You would also have to bootstrap the installation somehow. `curl` might be not present on the target, and so can `bash`. The package manager varierty is also quite gigantic. You would have to handle A LOT of platforms to bootstrap the Neovim installation.
5. Plugins are pretty ad-hoc. Some of them depend on some runtime (LSPs are a great example), some require packages to be installed on the host, some require fonts (basically every plugin that wants Nerd Fonts). You are usually expected to handle it yourself, but it doesn't mesh well with automation. And then, a lot of plugin managers support version pinning but don't enforce it by default, so you could easily get a different version of a plugin installed which would cause unexpected problems. And then some plugins also have their internal update mechanism (LSP servers, Treesitter), which can also explode on you.


# Better approach
1. We should use local Neovim for everything. That way we are sure that everything works, and it doesn't require extra effort.
2. We mount remote files locally, and open them in Neovim. SSHFS is one way to do it, but it's unmaintained and problematic. I suggest rclone - [it supports remote mounting](https://rclone.org/commands/rclone_mount/). It does a lot of what we'd like to have itself, it's reasonably efficient, it caches files locally (which means that if you experience a disconnect - you can continue your work) and it comes with a very nice perk of supporting a lot of cloud providers. Maybe you are storing JSONs or JS files on S3 - you can redact them with Neovim seamlessly! This is actually better than what VSCode can offer.
3. We'd have to sync the changes. Rclone has commands and settings for syncing changes: you could set it to sync changes every 5 seconds, and do it in background asynchronously. You would likely have to auto-write changes to local filesystem in Neovim just before that, which should be fine. This can be made reasonably efficient and nice to use if you play around with the settings, you'd be able to handle disconnects too. Rclone also lets you resolve conflicts!

# Prior work
https://github.com/ipod825/vim-netranger - implements a file manager that also can use `rclone`. This is basically what we want.
https://github.com/DanielWeidinger/nvim-sshfs - implementation with SSHFS

**Comment by jdrouhard** [comment]
The obvious glaring downside to the "local everything" approach is if you actually need to develop in the remote server's environment.

What if I have a work laptop with bare minimum essentials (a terminal emulator, maybe a neovim install, and a VPN), and I can't have the full development environment toolchain? Pick your poison: gcc with lots of of library dependencies, node with tons of npm modules, whatever. It might not even be the same OS. Windows->Linux, Linux->macOS, etc.

With vscode-remote, I can have a development experience that "feels" local whilst actually running in a completely different environment that is equipped appropriately. In this way, vscode-remote and remote nvim UI's approaches are superior to some sort of file synchronization method to pull files locally for editing.

**Comment by chaneyzorn** [comment]
Mounting remote files locally such as sshfs under poor network is not friendly for local git, and git operations become terribly  slow.

**Comment by ashwinvis** [comment]
I have played around with Neovim's `--remote-ui` option and some SSH tunneling of via sockets. I made a **semi-automated solution** here, https://codeberg.org/ashwinvis/ssh.nvim . Someone might find it useful

I agree particularly with @jdrouhard's take [[1](https://github.com/neovim/neovim/issues/21635#issuecomment-1374539337)]. What we have in neovim natively is:

- Communication over `scp://` using `netrw` which gives a **local everything** approach. Buffer is cached locally making it "snappy" even with high network latency, `:Lexplore` would show the remote files. The problem is, things like LSP, integrated `:terminal` and `:!` commands do not work.
- Using `nvim --remote-ui --server` or `neovide --server` option is a **remote everything** approach. It has works with some semi-automation as shown above, until someone makes a plugin to make this truly effortless and cross-platform. It does not have the downsides of `netrw` (LSP, terminal and commands works) but everything is synchronous. It works only with a good network and is painful to use over network with high latency.

People have mentioned VScode remote-ssh and Emacs TRAMP. Another tool which works well remotely out of the box is **JupyterLab**  because it was always built with a client-server model. The buffers are cached in the client's browser and everything is executed in the server.

**Comment by elijahmorg** [comment]
> curious if there would be a way to embed the distant library into neovim itself. It's a rust library, so if we exported it in a C-friendly way, it could be consumed.

I don't think this is a good idea. SSH is ubiquitous. Distant is very new and not currently supported.

The beauty of VCode remote is how seemless it is and it "just works" most of the time.

The security implications of SSH are also very well understood while distant would be something entirely new that users would have to understand.

I would be for this if 
A) It was seamless and the end user does not have to add distant to the remote machine
B) If all the communication could be wrapped in an SSH tunnel

otherwise I worry it would be too niche.

**Comment by chipsenkbeil** [comment]
> I would be for this if 
> A) It was seamless and the end user does not have to add distant to the remote machine
> B) If all the communication could be wrapped in an SSH tunnel
> 
> otherwise I worry it would be too niche.

Distant also supports acting as an ssh client, so you get both feature sets. So yes you can! :smile:

> The security implications of SSH are also very well understood while distant would be something entirely new that users would have to understand.

Fair enough. Distant uses a very modern encryption and authentication standard on part with ssh, but it definitely isn't battle tested or audited, although the crypto libraries have been audited. 


But advanced functionality like running a file watcher for remote changes, translating language server file references such that it can work on the remote machine, and remote filesystem searching are implemented using distant.

Regardless, using it as a client library would give you a high level abstraction for file io and process execution, rather than needing to work with ssh and the sftp subsystem.

As distant implements support for other remote platforms, neovim would also get it for free.

So I'm not necessarily sold on limiting to ssh. We've seen the wonders that Lua had over vim script! :smile:

**Comment by amitds1997** [comment]
I created a plugin for this use case: [**remote-nvim.nvim**](https://github.com/amitds1997/remote-nvim.nvim). If you use [devpod](https://github.com/loft-sh/devpod), you can also develop inside devcontainers using this plugin.

I deployed the first release of the plugin yesterday (**v0.0.1**) 🎉. Feel free to reach out to me if you have any questions about the plugin. You can drop your queries about the plugin over in the [release discussion](https://github.com/amitds1997/remote-nvim.nvim/discussions/30).

I have only tested this with MacOS and Linux. Support for Windows is planned but timelines are not decided yet.

**Comment by fecet** [comment]
> I created a plugin for this use case: [**remote-nvim.nvim**](https://github.com/amitds1997/remote-nvim.nvim). If you use [devpod](https://github.com/loft-sh/devpod), you can also develop inside devcontainers using this plugin.
> 
> I deployed the first release of the plugin yesterday (**v0.0.1**) tada. Feel free to reach out to me if you have any questions about the plugin. You can drop your queries about the plugin over in the [release discussion](https://github.com/amitds1997/remote-nvim.nvim/discussions/30).
> 
> I have only tested this with MacOS and Linux. Support for Windows is planned but timelines are not decided yet.

Does this do something magic or just setup neovim remotely and start a headless nvim client?

**Comment by amitds1997** [comment]
> Does this do something magic or just setup neovim remotely and start a headless nvim client?

No magic. It installs Neovim on the remote server in a custom folder and then starts a headless server over on remote and then launches a client locally.

**Comment by zeertzjq** [comment]
Oops

**Comment by CGamesPlay** [comment]
I want to add some brainstorming thoughts to this discussion to focus on what a real implementation of this for Neovim would actually look like. Continuing where @ashwinvis left off, what is the middle ground between "local everything" and "remote everything"? I'll call this the "hybrid remote editing" approach. The goal of this is to act like "remote everything", but where config is local and the UI isn't dependent on the network.

- All UI needs to be rendered locally. Buffers need to be available and edited locally. Swapfiles are local. This part looks just like using netrw.
- All filesystem access and process execution needs to happen remotely. Effectively, nvim should think that it's running on the remote system. This would mean that `!` filters would be remote, all language server integrations would be remote, `:terminal`, `:cd`, etc. Also, environment variables should all come from the remote host.
- For simplicity, I think the entire neovim process only exists in one editing context. To start a hybrid remote editing session, you would invoke nvim with a flag that enables it. This avoids a lot of headaches around mixing contexts, and is also how VS Code implements this.

How could we actually implement this? I think the best way is to give add a new concept of "location" to vim. The initial "location" is "default". While this location is active, all operations (FS, process execution, environment) work exactly as if they were performed on the remote machine.

There should be some APIs to change the location to "local", like `:local COMMAND` to run a command with a local location, a Lua API to use the local location for the duration of a function call, etc (I think a variety of these will be necessary). While the local location is set, operations work as they do today (e.g. they run in the context of the machine vim is running on).

The default location works like a sandbox that only has some specific escape hatches. This means that existing scripts that aren't location-aware can continue to work unaffected: as far as they can tell, they are running on the remote editing host. But new scripts will be able to "escape" with the local location (an example: opening a URL in the user's browser requires executing a process in the local location).

A quick "user story": Imagine starting `nvim --hybrid-remote-editing=myhost:path/to/folder`. Neovim starts as usual with your local init files, and `:pwd` reports that you are in `/home/user/path/to/folder`. If you `:e index.ts`, nvim finds and loads the appropriate `ftplugin` in the (local) `vimruntimepath`. Some script decides to run a linter, which will run on the remote host, resulting in some signs being applied to the buffer. When you run `:terminal`, the prompt is `user@myhost ~/path/to/folder` and it is running remotely. When you turn off wifi, you can continue to move around the buffer, but async jobs (like the linter) never finish and sync commands like `:w` hang nvim.

Open questions:
- What APIs are covered under the "location" concept? The goal of the default location is to effectively trick any user command or script that isn't "location-aware" that it's actually running on the remote system.
- What internal changes need to be made location-aware? Two obvious ones: init scripts and swap files.
- In this proposal, the `:source` builtin and all other script evaluation happens in the local location. This side-steps the security concerns, but it may be desirable to add a separate API to enable sourcing remote files.
- This proposal does not talk about installation of neovim on the remote at all. I think this is outside the scope of core. If Neovim provides an API to "restart in a hybrid remote editing context", then a script can implement this behavior itself.

**Comment by justinmk** [comment]
> Buffers need to be available and edited locally. Swapfiles are local.

Where did that requirement come from? 

> There should be some APIs to change the location to "local", like `:local COMMAND` to run a command with a local location,

This is too complicated and outside the scope of this issue. The goal of this issue is not full emacs TRAMP. The goal is just what I described in the main description: auto-download nvim on the remote and connecting to that from a local nvim, with a single command.

**Comment by CGamesPlay** [comment]
Sorry for being considered off-topic. I think it's worth pointing out that the subject of this issue says "like vscode remote-ssh", which does all editing locally and has the notion of local and remote execution contexts. That's why I thought it was related!

**Comment by fecet** [comment]
I personally strongly agree that neovim should have a remote mode as powerful as vscode. This is definitely a killer feature.

**Comment by mikew** [comment]
> Running Neovim on the remote takes CPU, RAM and disk space. This can be a complete dealbreaker for some of the more resource-intensive plugins, like rust-analyzer.

This is exactly what I'm after though. When Apple switched to ARM I wasn't affected, because all the work happens in a remote environment. Also means a small Macbook Air gets a lot of life, since nothing that intensive is actually running on it.

> As noted by @jdrouhard , latencies quickly become an issue, which makes remote development feeling no better than just SSHing into the host and typing manually.

You've listed this under problems of VSCode's approach, when it's actually not a problem with how they've architected things.

> We should use local Neovim for everything.

I disagree, reasoning stated above.

> We mount remote files locally, and open them in Neovim. SSHFS is one way to do it

I have never had any luck with SSHFS and it's counterparts on macOS.

**Comment by kellpossible** [comment]
Another reason to have work happening in the remote environment  (aside from more available RAM and CPU in a remote server than on a low end laptop) is when you have a poor internet connection, one that is good enough to maintain a text editing session, but not good enough for downloading large dependencies/containers required for development or uploading build results, etc.

**Comment by ivandimitrov8080** [comment]
It seems that we know what we want already and that thing is to offload the burden of running an IDE to a server so that your craptop can manage for a few more years.
How should one go about implementing this? I'm not a neovim dev I just use it and I like it so maybe some of you might know something obvious that I don't see. Since there's an option to have
> nvim --headless --listen 0.0.0.0:8080 and nvim --server=neovim.com:8080

then would it not make sense to have this work without latency by using caching? Why not?

The optimal solution would move the most amount of resource-intensive tasks to the server and leave the UI and only that on the client. Why can't neovim do that since it's already capable of being a server?

**Comment by justinmk** [comment]
Locking this since comments are now repeating things already discussed in detail above.
